Sign tsh.exe on tag builds

timothyb89 opened the pull request:
This adds a Makefile step to sign tsh.exe when the `$WINDOWS_SIGNING_CERT` env var is set to a base64-encoded pkcs12 code signing certificate. The certificate must not be password protected. This signing step replaces the previous Windows release step, which has been renamed to `release-windows-unsigned`. Unsigned builds generate artifacts with a `-unsigned` suffix; `release-windows` generates signed artifacts with backwards compatible names.

This includes a sample cert (`cert-dummy.pfx`) for CI pipeline testing. It should be removed in any eventual PR, along with the other modifications to the drone pipeline. The cert is imported into the environment in the `Makefile` for testing purposes; in practice it will be imported from a secure secret store (drone secrets, etc).

timothyb89 committed:
Sign tsh.exe on tag builds

This adds a Makefile step to sign tsh.exe when the
`$WINDOWS_SIGNING_CERTIFICATE` env var is set to a base64-encoded
pkcs12 code signing certificate. The certificate must not be password
protected.

This includes a sample cert (`cert-dummy.pfx`) for CI pipeline
testing. It should be removed in any eventual PR, along with the
other modifications to the drone pipeline. The cert is imported into
the environment in the `Makefile` for testing purposes; in practice
it will be imported from a secure secret store (drone secrets, etc).

timothyb89 committed:
Release v7.0.0-dev.38

timothyb89 committed:
Revert "Release v7.0.0-dev.38"

This reverts commit 45239c8923565bd353e712756b039bebc80c6fa6.

timothyb89 commented:
Just tested this with a dev tag: https://get.gravitational.com/teleport-v7.0.0-dev.38-windows-amd64-bin.zip

![Screen Shot 2021-08-13 at 4 39 57 PM](https://user-images.githubusercontent.com/3459954/129424610-12fe83eb-5c29-4b6f-a75b-58a5295d70ca.png)

webvictim commented on Makefile:
```diff
@@ -298,6 +298,29 @@ release-windows: clean all
 	mv teleport/tsh teleport/tsh.exe
+	# TODO: remove this
+	@WINDOWS_SIGNING_CERTIFICATE=$$(cat cert-dummy.pfx | base64); \
+	if [ ! -z "$${WINDOWS_SIGNING_CERTIFICATE}" ]; then \
```
I would be tempted to make this target fail safe and exit with error if the Windows signing certificate is not provided, so that we can't ever accidentally release an unsigned `tsh.exe` due to an empty/missing secret or a lack of access, etc. To make local development easier, this could potentially be gated to only exit when running in CI by checking whether the `DRONE` environment variable is set, similar to this: https://github.com/gravitational/teleport/blob/master/build.assets/Makefile#L49

Another alternative would be to do a similar thing to what we do with RPMs - have a `windows-unsigned` target which doesn't use a signature (for local development builds etc): https://github.com/gravitational/teleport/blob/master/Makefile#L687-L690

wadells replied:
I'd go even further and make signing a seperate make target (with Gus's fail safe property) that depends on the unsigned artifact.

timothyb89 replied:
I've now split this into `release-windows-unsigned` (which generates the usual .zip file with a `-unsigned` suffix) and `release-windows` which extracts the original .zip, signs + renames the exe, and rezips everything.

webvictim commented on dronegen/tag.go:
```diff
@@ -179,6 +179,7 @@ func tagPipeline(b buildType) pipeline {
 		"GOPATH":  value{raw: "/go"},
 		"OS":      value{raw: b.os},
 		"ARCH":    value{raw: b.arch},
+		//"WINDOWS_SIGNING_CERTIFICATE: value{fromSecret: "WINDOWS_SIGNING_CERTIFICATE"},
```
I realise this probably isn't complete yet, but you should gate this so it only exposes the secret when `b.os == "windows"`.

timothyb89 replied:
Good catch, I'll make the change

timothyb89 replied:
I've made the change, but for testing now also pass in the dummy cert via a raw drone env var.

timothyb89 committed:
Improve Windows code signing

 - Split signing into a separate step; `release-windows-unsigned` now
   performs the build, `release-windows` signs the binary.
 - Require `release-windows` to successfully generate a signed
   binary.
 - Clearly mark unsigned binaries and archives as such.
 - Guard against stdout secret leakage in Makefiles.
 - Move temporary cert data from Makefile into dronegen to test
   full pipeline.

timothyb89 committed:
Release v7.0.0-dev.39

wadells commented on Makefile:
```diff
@@ -286,19 +286,55 @@ release-unix: clean full
+	
+	@echo "---> Signing Windows binary."
+	@cert_path=$$(mktemp); \
+	echo "$${WINDOWS_SIGNING_CERT}" | base64 -d > "$$cert_path" || exit 1; \
```
I'd prefer we decode in the Drone logic, and have the Makefiles use a path to a file as input -- or rely on a wellknown (and .gitignored) path.  This keeps Make and Drone more loosely coupled. A base64 encoded file in an environment variable isn't a common build input -- we probably wouldn't use that same system with any other CI harness.

E.g. consider the case of testing locally. To me it feels obtuse to:

```console
WINDOWS_SIGNING_CERT="$(cat test.pfx | base 64)" make release-windows
```
As opposed to:

```console
WINDOWS_SIGNING_CERT=test.pfx make release-windows
```

timothyb89 replied:
I don't disagree, that would be a nice improvement (and more similar to how we handle gpg for signed Linux builds).

Unfortunately `make` [immediately changes directories](https://www.gnu.org/software/make/manual/make.html#Recursion) when we use `make -C build.assets ...` so any path contained within a `$WINDOWS_SIGNING_CERT` variable will be relative to an unknown base directory by the time we try to do anything with it. `make` apparently forgets its original cwd and provides no variables or functions to recover it, let alone resolve an absolute path from a user-provide env var.

We could expect anyone testing locally to provide an absolute path but IMO that's beginning to feel similarly obtuse.

timothyb89 replied:
Alright, it now uses a set file (`windows-signing-cert.pfx`) in the root of the source directory which sidesteps all the`make` issues and should be easier to test. It should be modestly harder to accidentally leak the secret as well.

wadells commented on Makefile:
```diff
@@ -286,19 +286,55 @@ release-unix: clean full
+# tsh.exe
+#
+.PHONY: release-windows
+release-windows: release-windows-unsigned
```
Depend on `$(RELEASE)-unsigned.zip` instead of the .PHONY `release-windows-unsigned` so that we don't force a clean and rebuild every time we sign. E.g:

```make
release-windows-unsigned: $(RELEASE)-unsigned.zip

$(RELEASE)-unsigned.zip: clean all
	@echo "---> Creating OSS release archive."
        ...

release-windows: $(RELEASE)-unsigned.zip
	@if [ -z "$${WINDOWS_SIGNING_CERT}" ]; then
        ...
```

wadells replied:
Never mind -- this won't work without also fixing `$(RELEASE)-unsigned.zip`'s dependencies, which is out of scope here.

timothyb89 committed:
Revert "Release v7.0.0-dev.39"

This reverts commit 32a2ee970c9f814592186c93b65ae0e3bc50ad7f.

timothyb89 commented:
Hmm, seems I broke drone with 32a2ee9, it just hangs forever and never starts the job. I guess maybe it doesn't work with `!!binary`, and the yaml serializer seems adamant about using that instead of a regular string. Oh well, I guess.

wadells commented:
The production `WINDOWS_SIGNING_CERT` secret is available in Drone now.

wadells commented:
> Hmm, seems I broke drone with [32a2ee9](https://github.com/gravitational/teleport/commit/32a2ee970c9f814592186c93b65ae0e3bc50ad7f), it just hangs forever and never starts the job. I guess maybe it doesn't work with `!!binary`, and the yaml serializer seems adamant about using that instead of a regular string. Oh well, I guess.

I wasn't able to find anything in the drone service or kubernetes runner logs about this.  If you continue with this testing strategy and see further trouble, ping me while the pod is running and I'll check it out.

timothyb89 committed:
Use an invalid cert string for testing purposes.

timothyb89 committed:
Release 7.0.0-dev.40

timothyb89 committed:
Revert "Release 7.0.0-dev.40"

This reverts commit 55c373bce108daa843c6fb8165c4ec952a747072.

timothyb89 commented:
My latest tagging test is failing appropriately so I'm at least confident the base64 env var -> `build.assets/Makefile` -> Docker buildbox -> `Makefile` pipeline is working properly (and I believe without any secret leakage from Drone).

timothyb89 committed:
Pass certs to the build process via a statically named file

Signed Windows builds now depend on a `.gitignore`'d
`windows-signing-cert.pfx` at the root of the source directory. This
should ease testing and help avoid accidental secret leakage.

timothyb89 committed:
Release 7.0.0-dev.41

timothyb89 committed:
Revert "Release 7.0.0-dev.41"

This reverts commit 03892f2d856e306131cc0a567a94352cf8eb3fb8.

timothyb89 committed:
Update .drone.yml

timothyb89 committed:
Release 7.0.0-dev.42

timothyb89 committed:
Revert "Release 7.0.0-dev.42"

This reverts commit 0ee09a2377ee65fd3e444c86bd21edb96bb28e1e.

timothyb89 committed:
Use production secret

webvictim commented on .drone.yml:
```diff
@@ -3643,6 +3643,7 @@ steps:
   - chown -R $UID:$GID /go
   - cd /go/src/github.com/gravitational/teleport
+  - echo "${WINDOWS_SIGNING_CERT}" | base64 -d > windows-signing-cert.pfx
   - make -C build.assets release-amd64
```
I would also add a line here to explicitly `rm -f windows-signing-cert.pfx` after the build is complete. This is good hygiene in general (deliberately removing key material as soon as it's been used) and will reduce the risk of any change to a subsequent stage accidentally uploading the cert.

We do actually use `/tmpfs` as an in-memory `EmptyDir` elsewhere in the pipeline to try and guarantee the ephemerality of key material and prevent things like this happening - see https://github.com/gravitational/teleport/blob/master/.drone.yml#L4329-L4331

I'm unsure whether this is actually improving security at all, though, so I'm not going to stipulate that it's required here.

timothyb89 replied:
That's a good thought, I've tweaked it to write the cert to a tmpfs and then symlink it into the build directory.

timothyb89 replied:
Walking that back, something about my implementation was not playing nice in Drone. I've opted not to waste more time debugging that and am now just removing the file following a successful build as you recommended.

timothyb89 committed:
Remove windows-signing-cert.pfx before continuing to the next step

Additionally, fix variable reference as the bracket syntax does not
seem to play nice with Drone.

timothyb89 commented:
Just rebased and retested: https://get.gravitational.com/teleport-v7.0.0-dev.47-windows-amd64-bin.zip

Testing against our new [security policy](https://goteleport.com/security/) I see:
```
PS C:\Users\Tim> Get-AuthenticodeSignature -FilePath .\tsh.exe


    Directory: C:\Users\Tim


SignerCertificate                         Status                                 Path
-----------------                         ------                                 ----
F2FBE7B8228122EB74DE2DC093DB81F8E6896253  Valid                                  tsh.exe


PS C:\Users\Tim>
PS C:\Users\Tim> (Get-AuthenticodeSignature -FilePath.\tsh.exe).SignerCertificate | Format-List


Subject      : CN=Gravitational Inc., O=Gravitational Inc., L=Oakland, S=California, C=US
Issuer       : CN=DigiCert SHA2 Assured ID Code Signing CA, OU=www.digicert.com, O=DigiCert Inc, C=US
Thumbprint   : F2FBE7B8228122EB74DE2DC093DB81F8E6896253
FriendlyName :
NotBefore    : 11/8/2020 5:00:00 PM
NotAfter     : 11/14/2023 4:59:59 PM
Extensions   : {System.Security.Cryptography.Oid, System.Security.Cryptography.Oid, System.Security.Cryptography.Oid,
               System.Security.Cryptography.Oid...}
```

... so it looks like everything's working as it should!

awly commented on .gitignore:
```diff
@@ -60,3 +60,6 @@ terraform.tfstate
 *.retry
+
+# Code signing certificate for Windows binaries
+windows-signing-cert.pfx
```
```suggestion
/windows-signing-cert.pfx
```

timothyb89 replied:
Oh, good catch, thanks!

timothyb89 committed:
Update .gitignore

Co-authored-by: Andrew Lytvynov <andrew@goteleport.com>

timothyb89 committed:
Merge branch 'master' into timothyb89/tsh-exe-signing-test

timothyb89 committed:
Merge branch 'master' into timothyb89/tsh-exe-signing-test
